DTLS 1.3: additions for event driven server in wolfssl-examples

julek-wolfssl opened the pull request:
- Add callback when we parse a verified ClientHello
- Allow the DTLS server to operate without maintaining state

julek-wolfssl committed:
Add callback when we parse a verified ClientHello

julek-wolfssl committed:
DTLS 1.3: allow the server to operate without maintaining state

dgarske reviewed (changes requested):
What is driving the addition of this client_hello good callback? Did you consider expanding the `WOLFSSL_CALLBACKS` and `HandShakeCallBack` used by `wolfSSL_accept_ex` / `wolfSSL_connect_ex` to support this feature?
If adding an API and callback function would be nice to not limit it to client_hello. This callback seems very DTLS v1.3 specific.

rizlik commented on src/tls.c:
```diff
@@ -6163,8 +6163,16 @@ static int TLSX_Cookie_Parse(WOLFSSL* ssl, const byte* input, word16 length,
-        return HRR_COOKIE_ERROR;
+    if (extension == NULL) {
+#ifdef WOLFSSL_DTLS13
+        if (ssl->options.dtls)
```
Is this for dtls 1.3 only? I think you should add `IsAtLeastv13` to the check

julek-wolfssl replied:
The cookie extension is only present in TLS 1.3 (I think) so you are right. I will add it.

rizlik reviewed (changes requested):
Is the CB supposed to work with DTLSv1.0 and v1.2 as well?

julek-wolfssl commented:
> Is the CB supposed to work with DTLSv1.0 and v1.2 as well?

It should but I haven't tested it yet.

> What is driving the addition of this client_hello good callback?

The motivation behind the new callback is to signal to the user that we have performed a successful return-routability test with the current peer. This means that the current object should dedicated to only serving this connection and a new object should be created to listen for new connections. I will see if I can make use of the existing handshake callbacks instead of writing a new callback. I agree that the callback is very DTLS specific.

julek-wolfssl commented on src/internal.c:
```diff
@@ -8783,6 +8783,14 @@ static int EdDSA_Update(WOLFSSL* ssl, const byte* data, int sz)
+    byte digest[WC_MAX_DIGEST_SIZE];
+
+    WOLFSSL_MSG("HashRaw:");
+    WOLFSSL_MSG("Data:");
```
The DEBUG_TLS logs are only useful for dev work already so I think it makes sense to include as much data as possible. This also helps to figure out where we went wrong with the hash.

rizlik replied:
Besides dev work logs are often asked for debugging non-local or not easily reproducible systems

julek-wolfssl replied:
In that case I think its best to have all the info available ASAP.

julek-wolfssl commented on src/tls.c:
```diff
@@ -6163,8 +6163,16 @@ static int TLSX_Cookie_Parse(WOLFSSL* ssl, const byte* input, word16 length,
+    if (extension == NULL) {
+#ifdef WOLFSSL_DTLS13
+        if (ssl->options.dtls)
+            /* TODO: Should we allow a ClientHello with a valid cookie even if
```
I think we should because this might happen in multi-thread load balancing. A different thread replied earlier but now we got the CH2 so I don't think this should be an error.

julek-wolfssl commented:
@dgarske I took a look at the `WOLFSSL_CALLBACKS` handshake callbacks but I don't think that this is right solution here. These callbacks only inform the user what packets have been processed and the `hsCb` callback only gets called after the handshake finishes. There is also the `protoMsgCb` callback but that is gated behind `OPENSSL_EXTRA` and it still doesn't provide any information if the ClientHello is verified or not. This will not work for DTLS because we want to isolate the socket as soon as we have processed a verified ClientHello. For this I think that a dedicated callback for DTLS is reasonable.

julek-wolfssl committed:
Address code review

julek-wolfssl committed:
Add docs for new features

julek-wolfssl committed:
DTLS 1.3: tie cookie to peer address

julek-wolfssl committed:
dtls: wolfSSL_set_dtls_fd_connected

wolfSSL_set_dtls_fd_connected sets the connected socket file descriptor. This descriptor should be called without addr and addr_len.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -1183,6 +1183,14 @@ int wolfSSL_dtls_set_peer(WOLFSSL* ssl, void* peer, unsigned int peerSz)
     if (ssl == NULL)
         return WOLFSSL_FAILURE;
 
+    if (peer == NULL || peerSz == 0) {
```
Added docs.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -12745,6 +12765,20 @@ int wolfSSL_DTLS_SetCookieSecret(WOLFSSL* ssl,
 
 #endif /* NO_WOLFSSL_SERVER */
 
+#ifdef WOLFSSL_DTLS
```
Wrapped.

dgarske commented on src/wolfio.c:
```diff
@@ -382,11 +382,35 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+            else
+                dtlsCtx->peer.bufSz = 0;
+        }
+        peer = dtlsCtx->peer.sa;
```
```
./configure '--enable-all' '--enable-sp-math-all' 'CC=g++' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK' && make
...
src/wolfio.c:410:30: error: invalid conversion from ‘void*’ to ‘SOCKADDR_S*’ {aka ‘sockaddr_storage*’} [-fpermissive]
  410 |         peer = dtlsCtx->peer.sa;
      |                ~~~~~~~~~~~~~~^~
      |                              |
      |                              void*
```

julek-wolfssl replied:
Fixed cast.

dgarske commented on wolfssl/internal.h:
```diff
@@ -4460,6 +4463,11 @@ struct WOLFSSL {
 #ifdef WOLFSSL_STATIC_MEMORY
     WOLFSSL_HEAP_HINT heap_hint;
 #endif
+#ifdef WOLFSSL_DTLS
```
Do we always need this DTLS feature? Perhaps we can add a new macro for it? Shouldn't it be disabled if building with `NO_WOLFSSL_SERVER`? I want to make sure embedded DTLS customers with small flash/ram don't have issues.

julek-wolfssl replied:
I'm not sure putting it behind a new define makes sense. Adding a `!NO_WOLFSSL_SERVER` is a good idea since this is only useful for a server.

dgarske commented on wolfssl/internal.h:
```diff
@@ -2201,13 +2201,16 @@ WOLFSSL_LOCAL int DoVerifyCallback(WOLFSSL_CERT_MANAGER* cm, WOLFSSL* ssl,
     int rfd;
     int wfd;
+    byte userSet:1;
+    byte connected:1; /* Set when the rfd and wfd are connected sockets */
```
Can we come up with a better name for this member? It seems to only change the use of the "peer" sockaddr in the recvfrom and sendto.

julek-wolfssl replied:
I honestly can't think of a better name that won't turn into a Java-esque sentence. When `connected` is set, this indicates that the sockets in `rfd` and `wfd` are "connected" (`connect()` was called on them in addition to `bind`). This means that `sendto` and `recvfrom` don't need to specify and store the peer address. Also if I remember correctly, @ejohnstown reported on another PR that using `sendto` on a connected socket on a Mac resulted on a failure. This may require us to switch to using just `send` in `EmbedSendTo` on a connected socket.

In short: can you think of a better name to describe this behaviour because I can't?

rizlik replied:
maybe `connectedSocket` or `connectedDatagramSocket`

julek-wolfssl replied:
I don't think that is a better name. Its too verbose and `connected` in the context of the `WOLFSSL_DTLS_CTX` object already only applies to the socket.

julek-wolfssl committed:
Apply connected to sendto and address code review

julek-wolfssl commented on wolfssl/ssl.h:
```diff
@@ -1045,6 +1045,7 @@ WOLFSSL_API int wolfSSL_CTX_set1_param(WOLFSSL_CTX* ctx, WOLFSSL_X509_VERIFY_PAR
 WOLFSSL_API int  wolfSSL_is_server(WOLFSSL* ssl);
 WOLFSSL_API WOLFSSL* wolfSSL_write_dup(WOLFSSL* ssl);
 WOLFSSL_ABI WOLFSSL_API int  wolfSSL_set_fd(WOLFSSL* ssl, int fd);
+WOLFSSL_API int wolfSSL_set_dtls_fd_connected(WOLFSSL* ssl, int fd);
```
Wrapped.

ejohnstown reviewed (changes requested):
When I tested this locally, merged with master (0459e83), the unit.test would fail:
```
starting DTLSv1.3 suite - downgrade
trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
SSL_accept error -326, record layer version error
wolfSSL error: SSL_accept failed
```
I used the configuration:
```
john:~/src/wolfssl (julek-wolfssl-dtls-good-ch-cb) % ./config.status --config
--enable-dtls --enable-dtls13 --enable-hrrcookie
```

dgarske committed:
Rename callback to `wolfDTLS_SetChGoodCb` and add doxygen for it. Clarify `DTLS_CTX.connected`. Fix build errors for `./configure --enable-dtls --enable-dtls13 --disable-examples CFLAGS="-DNO_WOLFSSL_SERVER"`.

dgarske reviewed (approved):
Renamed the function to DTLS specific. Did not try and run the code. John did and he's having issues (see his comment). I will wait to merge until those are resolved.

rizlik commented:
> When I tested this locally, merged with master ([0459e83](https://github.com/wolfSSL/wolfssl/commit/0459e83a59d83ac7ebba3271a09ec26c52a56642)), the unit.test would fail:
> 
> ```
> starting DTLSv1.3 suite - downgrade
> trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
> trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
> SSL_accept error -326, record layer version error
> wolfSSL error: SSL_accept failed
> ```
This problem is on master as well and it's not specific to this PR (we don't have any DTLS1.3 configuration in the automatic test yet). I'm going to track down and fix this.

rizlik commented:
> > When I tested this locally, merged with master ([0459e83](https://github.com/wolfSSL/wolfssl/commit/0459e83a59d83ac7ebba3271a09ec26c52a56642)), the unit.test would fail:
> > ```
> > starting DTLSv1.3 suite - downgrade
> > trying server command line[3571]: SuiteTest -vd -7 2 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -2 -p 0 
> > trying client command line[3572]: SuiteTest -v 3 -u -l TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 -f -2 -p 51930 
> > SSL_accept error -326, record layer version error
> > wolfSSL error: SSL_accept failed
> > ```
> 
> This problem is on master as well and it's not specific to this PR (we don't have any DTLS1.3 configuration in the automatic test yet). I'm going to track down and fix this.

Indeed this was introduced by 060dfe1a693940fff31616570166be7f53b49ad8.

[#5315 ](https://github.com/wolfSSL/wolfssl/pull/5315) should fix this

rizlik commented on doc/dox_comments/header_files/ssl.h:
```diff
@@ -1631,9 +1631,40 @@ int  wolfSSL_set_fd (WOLFSSL* ssl, int fd);
+/*!
+    \ingroup Setup
+
+    \brief Allows setting a callback for DTLS client hello "good".
```
it should be mentioned that if HRR Cookie is enabled, this CB is triggered only for CH with a valid cookie.

julek-wolfssl replied:
Added description.

julek-wolfssl commented:
@dgarske I can confirm that https://github.com/wolfSSL/wolfssl/pull/5315 solves @ejohnstown's issue.

julek-wolfssl committed:
Use wc_HmacInit and wc_HmacFree in cookie logic

julek-wolfssl committed:
Reset cookie when resetting DTLS 1.3 state

julek-wolfssl commented:
Rebased onto master and force pushed.

julek-wolfssl committed:
Expand wolfDTLS_SetChGoodCb() docs

ejohnstown reviewed (approved):
The issue I was having was definitely outside of this PR.
